ui/raidboss/data/04-sb/ultimate/unending_coil_ultimate.ts

```diff
@@ -4,16 +4,203 @@ import Outputs from '../../../../../resources/outputs';
 import { Responses } from '../../../../../resources/responses';
 import Util from '../../../../../resources/util';
 import ZoneId from '../../../../../resources/zone_id';
+import { RaidbossData } from '../../../../../types/data';
+import { TriggerSet } from '../../../../../types/trigger';
+
+export interface Data extends RaidbossData {
+  // TODO: replace partyList with data.party
+  partyList: { [name: string]: boolean };
+  hpThresholds: number[];
+  monitoringHP: boolean;
+  currentPhase: number;
+  hatch?: string[];
+  doomCount?: number;
+  dooms?: { [doomIdx: number]: string | null };
+  fireDebuff: boolean;
+  iceDebuff: boolean;
+  naelFireballCount: number;
+  fireballs: { [fireballIdx: number]: string[] };
+  tookThreeFireballs?: boolean;
+  seenDragon: { [dragonName: string]: boolean };
+  // naelDragons[direction 0-7 (N-NW)] => boolean
+  naelDragons: number[];
+  naelMarks?: string[];
+  calledNaelDragons: boolean;
+  wideThirdDive: boolean;
+  unsafeThirdMark: boolean;
+  naelDiveMarkerCount: number;
+  trio?: string;
+  shakers: string[];
+  megaStack: string[];
+  octetMarker: string[];
+  lastOctetMarker?: string;
+  exaflareCount: number;
+  akhMornCount: number;
+  mornAfahCount: number;
+}
+
+const resetTrio = (data: Data, trio: string) => {
+  data.trio = trio;
+  data.shakers = [];
+  data.megaStack = [];
+};
+
+// Begin copy and paste from dragon_test.js.
+const modDistance = (mark: number, dragon: number) => {
+  const oneWay = (dragon - mark + 8) % 8;
+  const otherWay = (mark - dragon + 8) % 8;
+  const distance = Math.min(oneWay, otherWay);
+  console.assert(distance >= 0);
+  return distance;
+};
+
+const badSpots = (mark: number, dragon: number) => {
+  // All spots between mark and dragon are bad.  If distance == 1,
+  // then the dragon hits the spot behind the mark too.  e.g. N
+  // mark, NE dragon will also hit NW.
+  const bad = [];
+  const distance = modDistance(mark, dragon);
+  console.assert(distance > 0);
+  console.assert(distance <= 2);
+  if ((mark + distance + 8) % 8 === dragon) {
+    // Clockwise.
+    for (let i = 0; i <= distance; ++i)
+      bad.push((mark + i) % 8);
+    if (distance === 1)
+      bad.push((mark - 1 + 8) % 8);
+  } else {
+    // Widdershins.
+    for (let i = 0; i <= distance; ++i)
+      bad.push((mark - i + 8) % 8);
+    if (distance === 1)
+      bad.push((mark + 1) % 8);
+  }
+  return bad;
+};
+
+const findDragonMarks = (array: number[]):
+    undefined | { wideThirdDive: boolean; unsafeThirdMark: boolean; marks: number[] } => {
+  const marks = [-1, -1, -1];
+  let isWideThirdDive = false;
+
+  const dragons = [];
+  for (let i = 0; i < 8; ++i) {
+    if (array[i])
+      dragons.push(i);
+  }
+
+  if (dragons.length !== 5)
+    return;
+
+  const [d0, d1, d2, d3, d4] = dragons;
```

Comment:
(I wonder if TypeScript is smart enough for:

```ts
if (dragons.some(dragon => dragon === undefined))
  return;
const [d0, d1, d2, d3, d4] = dragons;
```
Or, if this was you covering for undefineds, I think this looks better? Feel free to disagree!)

Reply:
Yeah, this is not enough to convince TypeScript.  TypeScript has no knowledge of the length of this array, and so it could be of length 0, containing no undefineds, and yet `d0` etc would all be undefined.

However also, TypeScript is not smart enough to follow this logic either:
```typescript
  if ([d0, d1, d2, d3, d4].some((dragon) => dragon === undefined))
    return;
```

I think yours looks a lot better too.  We could arguably add a type guard function if this came up a lot, but I think better to just be explicit here.